helix-core/src/selection.rs

```diff
@@ -1,30 +1,50 @@
-//! Selections are the primary editing construct. Even a single cursor is defined as an empty
-//! single selection range.
+//! Selections are the primary editing construct. Even a single cursor is
+//! defined as a single empty or 1-wide selection range.
 //!
 //! All positioning is done via `char` offsets into the buffer.
-use crate::{Assoc, ChangeSet, RopeSlice};
+use crate::{
+    graphemes::{
+        ensure_grapheme_boundary_next, ensure_grapheme_boundary_prev, next_grapheme_boundary,
+        prev_grapheme_boundary,
+    },
+    Assoc, ChangeSet, RopeSlice,
+};
 use smallvec::{smallvec, SmallVec};
 use std::borrow::Cow;
 
-#[inline]
-fn abs_difference(x: usize, y: usize) -> usize {
-    if x < y {
-        y - x
-    } else {
-        x - y
-    }
-}
-
-/// A single selection range. Anchor-inclusive, head-exclusive.
+/// A single selection range.
```

Comment:
I would mention that the positions represent gap indices, signifying a gap between two characters. I also like this diagram from the tests:

https://github.com/helix-editor/helix/blob/f96b8b769b3c7457935b5c02db870af97036f7b6/helix-view/src/document.rs#L1130-L1135

Reply:
Oh, weird.  I could have sworn I put that in the docs.  Total goof on my part.  Thanks for catching that!